File: 18_Bit_Manipulation/06_Sum_of_Two_Integers/0371-sum-of-two-integers.py
"""
Problem: LeetCode 371 - Sum of Two Integers

Key Idea:
To calculate the sum of two integers without using the + and - operators, we can use bitwise operations. We use the XOR operation (^) to calculate the sum without considering the carry, and the AND operation (&) followed by a left shift (<<) to calculate the carry. We repeat these steps until there is no carry left.

Time Complexity:
- The number of iterations required to calculate the sum is proportional to the number of bits in the input integers, which is constant. Therefore, the time complexity is O(1).

Space Complexity:
- The space complexity is O(1), as we use a constant amount of space.
"""


class Solution:
    def getSum(self, a: int, b: int) -> int:
        MASK = 0xFFFFFFFF
        MAX_INT = 0x7FFFFFFF

        while b != 0:
            a, b = (a ^ b) & MASK, ((a & b) << 1) & MASK

        return a if a <= MAX_INT else ~(a ^ MASK)

File: 18_Bit_Manipulation/06_Sum_of_Two_Integers/0371-sum-of-two-integers.rs
/*
Problem: LeetCode 371 - Sum of Two Integers

Key Idea:
The key idea is to use bitwise operations to perform addition without using the `+` operator. We can use bitwise XOR to perform the addition without carrying, and bitwise AND to calculate the carry.

Approach:
1. Initialize two variables, 'a' and 'b', to represent the two integers to be added.
2. Loop while 'b' is not zero:
   a. Calculate the sum without carrying by performing 'a XOR b' and store it in 'a'.
   b. Calculate the carry by performing '(a AND b) << 1' and store it in 'b'.
3. After the loop, 'a' will contain the result, which is the sum of the two integers.

Time Complexity:
O(1), as the number of iterations in the loop is constant (32 bits for integers).

Space Complexity:
O(1), as we use only a constant amount of extra space.
*/

impl Solution {
    pub fn get_sum(a: i32, b: i32) -> i32 {
        let mut x = a;
        let mut y = b;

        while y != 0 {
            let carry = x & y;
            x = x ^ y;
            y = carry << 1;
        }

        x
    }
}

/*
impl Solution {
    pub fn get_sum(a: i32, b: i32) -> i32 {
        return [a, b].iter().sum();
    }
}
*/
